### PR TITLE
expression: implementations of MySQL load_file function in TiDB

### PR DESCRIPTION
**Pull request by cfzjywxk**
### What problem does this PR solve? 

implementations of MySQL load_file function in TiDB

ref https://github.com/pingcap/tidb/issues/8436

### What is changed and how it works?

implementations of MySQL load_file function in TiDB

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change
 - Has interface methods change

Side effects

Related changes

 - Need to update the documentation
 - Need to be included in the release note

### CODE REVIEW

File: expression/builtin_string.go [hunk 1]
```diff
@@ -3491,5 +3493,67 @@ type loadFileFunctionClass struct {
...
+}
+
+func (b *builtinInstrLoadFileSig) Clone() builtinFunc {
+	newSig := &builtinInstrLoadFileSig{}
+	newSig.cloneFrom(&b.baseBuiltinFunc)
+	newSig.maxAllowedPacket = b.maxAllowedPacket
+	return newSig
+}
+
+//MySQL implementation return NULL value when error occurs
```
Comment (jackysp) [inline expression/builtin_string.go]: It is better to add a space at the start of the comment.

**Comment by jackysp** [comment]
It seems the commit is not from your GitHub ID, the CLA check failed.

File: expression/builtin_string.go [hunk 2]
```diff
@@ -3491,5 +3493,67 @@ type loadFileFunctionClass struct {
...
+func (b *builtinInstrLoadFileSig) evalString(row chunk.Row) (string, bool, error) {
+	filePath, isNull, err := b.args[0].EvalString(b.ctx, row)
+	if isNull || err != nil {
+		return "", true, nil
+	}
+	//TODO: check filePath under secure_file_path(not exists in TiDB now)
```
Comment (lysu) [inline expression/builtin_string.go]: - - maybe we should ship this function after secure_file_path done?

it's too dangerous for someone can do `SELECT LOAD_FILE('/etc/passwd');`

Comment (cfzjywxk) [reply expression/builtin_string.go]: here besides the secure_file_path is needed as well as privlige control,   In MySQL SUPER FILE is needed to execute this function, this is another TODO @lysu

File: expression/builtin_string.go [hunk 3]
```diff
@@ -3491,5 +3493,67 @@ type loadFileFunctionClass struct {
 }
 
 func (c *loadFileFunctionClass) getFunction(ctx sessionctx.Context, args []Expression) (builtinFunc, error) {
-	return nil, errFunctionNotExists.GenWithStackByArgs("FUNCTION", "load_file")
+	if err := c.verifyArgs(args); err != nil {
+		return nil, err
+	}
+	valStr, _ := ctx.GetSessionVars().GetSystemVar(variable.MaxAllowedPacket)
+	maxAllowedPacket, err := strconv.ParseInt(valStr, 10, 64)
```
Comment (zz-jason) [inline expression/builtin_string.go]: According to the MySQL document [LOAD_FILE(file_name)](https://dev.mysql.com/doc/refman/5.7/en/string-functions.html#function_load-file), we should also check whether the user has the `FILE` privilege:

> you must have the FILE privilege.

Comment (cfzjywxk) [reply expression/builtin_string.go]: @zz-jason  MySQL has two checks "file_secure_path" and "FILE privilege" , both seems not exists in TiDB now, maybe it's better to release this func after more compicated privilege system

Comment (zz-jason) [reply expression/builtin_string.go]: > MySQL has two checks "file_secure_path" and "FILE privilege" , both seems not exists in TiDB now

@tiancaiamao Could you help @cfzjywxk to add the `FILE` privilege?

File: expression/builtin_string.go (hunk 1 above)
Comment (tiancaiamao) [reply expression/builtin_string.go]: The comment for this function should begin with `evalString` ...
https://golang.org/doc/effective_go.html#commentary

> Doc comments work best as complete sentences, which allow a wide variety of automated presentations. The first sentence should be a one-sentence summary that starts with the name being declared.

File: expression/builtin_string.go (hunk 2 above)
Comment (tiancaiamao) [reply expression/builtin_string.go]: It's a security vulnerability if we simply add this function.
An attacker can get any file he wants in on the host os.

File: expression/builtin_string.go (hunk 3 above)
Comment (cfzjywxk) [reply expression/builtin_string.go]: priv system is another task, is there a detailed design or implementation plan or roadmap about privilege system in TiDB ?   or we just make it full compatible with MySQL privilege system?
@zz-jason @tiancaiamao

Comment (zz-jason) [reply expression/builtin_string.go]: Yeah, we can do it in another PR. Just to mention @tiancaiamao to notice that we haven't implemented the `FILE` privilege. 😄

File: expression/builtin_string.go (hunk 1 above)
Comment (zz-jason) [reply expression/builtin_string.go]: @cfzjywxk PTAL at this comment.

File: expression/builtin_string.go (hunk 2 above)
Comment (zz-jason) [reply expression/builtin_string.go]: @cfzjywxk Could you file an issue about this TODO?

File: expression/builtin_string.go (hunk 3 above)
Comment (zz-jason) [reply expression/builtin_string.go]: @cfzjywxk How about filing an issue about this improvement?

**Comment by zz-jason** [comment]
@cfzjywxk friendly ping, any update?

**Comment by cfzjywxk** [comment]
> @cfzjywxk friendly ping, any update?

close it, reopen if needed in the future

**Pull request closed without merging**